### PR TITLE
Fix NPE when the plugin closes

### LINKED ISSUE #16404: DLQ throw NPE when the pipeline shutdown
Pipeline running dlq-input throws NullPointerException when received SIGTERM with the following conditions
- dead letter queue directory has never had a available segment
- commit_offsets => true 
- start_timestamp is not specified
- sincedb_path => "/path/to/sincedb" 

When DeadLetterQueueReader finds [no available](https://github.com/logstash-plugins/logstash-input-dead_letter_queue/blob/main/src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java#L125) segment in the path
- currentReader ([segment](https://github.com/elastic/logstash/blob/v8.15.0/logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java#L92)) is unset (null) [here](https://github.com/elastic/logstash/blob/v8.15.0/logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java#L345-L370)
- [readerHasState](https://github.com/logstash-plugins/logstash-input-dead_letter_queue/blob/main/src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java#L126) is set to `true`.

Logstash received SIGTERM,  close() try to update [SinceDB](https://github.com/logstash-plugins/logstash-input-dead_letter_queue/blob/v2.0.0/src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java#L159) before quit by getting the [path](https://github.com/elastic/logstash/blob/v8.15.0/logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java#L488) of segment

### Steps

```
# dlq.conf
input {
    dead_letter_queue {
        path => "/tmp/logstash/data/dead_letter_queue"  # /tmp/logstash/data/dead_letter_queue/pipe1 should be empty folder
        sincedb_path => "/path/to/sincedb" # a valid non empty sincedb
        commit_offsets => true
        pipeline_id => "pipe1"
    }
}
output {
    stdout {  }
}
```
- Run `bin/logstash -f dlq.conf`
- After pipeline started, shutdown logstash
- You can see the error log

### Logs

```
[2024-08-21T08:11:24,812][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::StopAndDelete/pipeline_id:deadletterqueue-standalone, :exception=>"Java::JavaLang::NullPointerException", :message=>"Cannot invoke \"org.logstash.common.io.RecordIOReader.getPath()\" because \"this.currentReader\" is null", :backtrace=>["org.logstash.common.io.DeadLetterQueueReader.getCurrentSegment(DeadLetterQueueReader.java:428)", "org.logstash.input.SinceDB.getUpdated(SinceDB.java:76)", "org.logstash.input.DeadLetterQueueInputPlugin.close(DeadLetterQueueInputPlugin.java:159)", "usr.share.logstash.vendor.bundle.jruby.$3_dot_1_dot_0.gems.logstash_minus_input_minus_dead_letter_queue_minus_2_dot_0_dot_0.lib.logstash.inputs.dead_letter_queue.RUBY$method$stop$0(/usr/share/logstash/vendor/bundle/jruby/3.1.0/gems/logstash-input-dead_letter_queue-2.0.0/lib/logstash/inputs/dead_letter_queue.rb:107)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:152)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:148)", "org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)", "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:456)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:195)", "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:350)", "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:128)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:115)", "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:446)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:92)", "org.jruby.RubySymbol$SymbolProcBody.yieldInner(RubySymbol.java:1508)", "org.jruby.RubySymbol$SymbolProcBody.doYield(RubySymbol.java:1523)", "org.jruby.runtime.BlockBody.yield(BlockBody.java:108)", "org.jruby.runtime.Block.yield(Block.java:189)", "org.jruby.RubyArray.each(RubyArray.java:1989)", "org.jruby.RubyArray$INVOKER$i$0$0$each.call(RubyArray$INVOKER$i$0$0$each.gen)", "org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroBlock.call(JavaMethod.java:561)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:90)", "org.jruby.ir.instructions.CallBase.interpret(CallBase.java:548)", "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:367)", "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66)", "org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:76)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:164)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:151)", "org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)", "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:456)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:195)", "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:350)", "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:128)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:115)", "org.jruby.ir.targets.indy.InvokeSite.fail(InvokeSite.java:777)", "org.jruby.ir.targets.indy.InvokeSite.fail(InvokeSite.java:809)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline_action.stop_and_delete.RUBY$block$execute$1(/usr/share/logstash/logstash-core/lib/logstash/pipeline_action/stop_and_delete.rb:30)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipelines_registry.RUBY$method$terminate_pipeline$0(/usr/share/logstash/logstash-core/lib/logstash/pipelines_registry.rb:192)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipelines_registry.RUBY$method$terminate_pipeline$0$__VARARGS__(/usr/share/logstash/logstash-core/lib/logstash/pipelines_registry.rb:184)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:139)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:112)", "org.jruby.ir.targets.indy.InvokeSite.performIndirectCall(InvokeSite.java:725)", "org.jruby.ir.targets.indy.InvokeSite.invoke(InvokeSite.java:657)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline_action.stop_and_delete.RUBY$method$execute$0(/usr/share/logstash/logstash-core/lib/logstash/pipeline_action/stop_and_delete.rb:29)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline_action.stop_and_delete.RUBY$method$execute$0$__VARARGS__(/usr/share/logstash/logstash-core/lib/logstash/pipeline_action/stop_and_delete.rb:28)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:139)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:112)", "org.jruby.ir.targets.indy.InvokeSite.performIndirectCall(InvokeSite.java:735)", "org.jruby.ir.targets.indy.InvokeSite.invoke(InvokeSite.java:657)", "usr.share.logstash.logstash_minus_core.lib.logstash.agent.RUBY$block$converge_state$2(/usr/share/logstash/logstash-core/lib/logstash/agent.rb:386)", "org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:141)", "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:64)", "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)", "org.jruby.runtime.Block.call(Block.java:144)", "org.jruby.RubyProc.call(RubyProc.java:352)", "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:111)", "java.base/java.lang.Thread.run(Thread.java:840)"]}
```

### PR DESCRIPTION
**Pull request by kaisecheng**
Fix NPE when the plugin closes with the following conditions
- dead letter queue directory has never had a available segment
- commit_offsets => true
- start_timestamp is not specified
- sincedb_path => "/path/to/sincedb"

This commit catches NullPointerException when updating SinceDB before shutdown

Fix: https://github.com/elastic/logstash/issues/16404

---

To test this PR 
```
# dlq.conf
input {
    dead_letter_queue {
        path => "/tmp/logstash/data/dead_letter_queue"  # /tmp/logstash/data/dead_letter_queue/pipe1 should be empty folder
        sincedb_path => "/path/to/sincedb" # a valid non empty sincedb
        commit_offsets => true
        pipeline_id => "pipe1"
    }
}
output {
    stdout {  }
}
```
- Run `bin/logstash -f dlq.conf`
- After pipeline started, shutdown logstash
- You should see NO error

### CODE REVIEW

File: src/test/java/org/logstash/input/DeadLetterQueueInputPluginTests.java [hunk 1]
```diff
@@ -129,6 +130,21 @@ public void testTimestamp() throws Exception {
...
+        SinceDB sinceDB = new SinceDB.AssignedDB(sincePath, garbageSegment, 0);
+        sinceDB.flush();
+        // Delete segment in DLQ folder
+        Files.delete(garbageSegment);
+
+        DeadLetterQueueInputPlugin plugin = new DeadLetterQueueInputPlugin(dir, true, sincePath, null, false, metricsSink);
+
+        plugin.register();
+        plugin.close();
+    }
```
Comment (andsel) [inline src/test/java/org/logstash/input/DeadLetterQueueInputPluginTests.java:147]:
```suggestion
    }

```

File: src/main/java/org/logstash/input/SinceDB.java [hunk 2]
```diff
@@ -76,6 +76,22 @@ static SinceDB getUpdated(SinceDB oldSinceDb, DeadLetterQueueReader reader) {
...
+    /**
+     * Attempts to update the SinceDB. This operation may throw a NullPointerException
+     * if DeadLetterQueueReader.currentReader is null, which can occur if
+     * DeadLetterQueueReader.segments has never had an available segment.
+     * The exception is caught and ignored.
+     * @param oldSinceDb
+     * @param reader
+     * @return new instance of SinceDB or UnassignedDB
+     */
+    static SinceDB mayUpdate(SinceDB oldSinceDb, DeadLetterQueueReader reader) {
```
Comment (andsel) [inline src/main/java/org/logstash/input/SinceDB.java]:
```suggestion
    static SinceDB tryUpdate(SinceDB oldSinceDb, DeadLetterQueueReader reader) {
```

File: src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java [hunk 3]
```diff
@@ -156,7 +156,7 @@ public void close() {
         final DeadLetterQueueReader queueReader = this.queueReader;
         if (queueReader != null && commitOffsets && readerHasState.get()) {
             logger.debug("retrieving current DLQ segment and position");
-            sinceDb = SinceDB.getUpdated(sinceDb, queueReader);
+            sinceDb = SinceDB.mayUpdate(sinceDb, queueReader);
```
Comment (andsel) [inline src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java]:
```suggestion
            sinceDb = SinceDB.tryUpdate(sinceDb, queueReader);
```

File: src/main/java/org/logstash/input/SinceDB.java [hunk 4]
```diff
@@ -76,6 +76,22 @@ static SinceDB getUpdated(SinceDB oldSinceDb, DeadLetterQueueReader reader) {
...
     }
 
+    /**
+     * Attempts to update the SinceDB. This operation may throw a NullPointerException
+     * if DeadLetterQueueReader.currentReader is null, which can occur if
+     * DeadLetterQueueReader.segments has never had an available segment.
+     * The exception is caught and ignored.
+     * @param oldSinceDb
+     * @param reader
+     * @return new instance of SinceDB or UnassignedDB
```
Comment (andsel) [inline src/main/java/org/logstash/input/SinceDB.java]:
```suggestion
     *
     * @param oldSinceDb of SinceDB to update
     * @param reader the reader that needs to be queried.
     * @return new instance of SinceDB or UnassignedDB if the reader doesn't point to an existing segment.
```

**Review by andsel — CHANGES_REQUESTED** [review]
The PR works as declared protecting against empty DLQ paths.
Left some suggestions on better naming.

**Comment by kaisecheng** [comment]
@andsel thanks for the review and suggestion. 

I am thinking if [readerHasState](https://github.com/logstash-plugins/logstash-input-dead_letter_queue/blob/v2.0.0/src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java#L126) should set to `false` when `queueReader.setCurrentReaderAndPosition()` find no segment, hence, [SinceDB](https://github.com/logstash-plugins/logstash-input-dead_letter_queue/blob/v2.0.0/src/main/java/org/logstash/input/DeadLetterQueueInputPlugin.java#L157-L160) no need to update when close. The issue seems to be misalignment of the state rather than missing try-catch. WDYT?

Edit: `queueReader.setCurrentReaderAndPosition()` doesn't expose the status of `queueReader.currentReader`. Without changing the core, now the NPE can only try-catch at close phase

**Comment by andsel** [comment]
@kaisecheng I agree with you, if the reader doesn't have a link to a valid segment (`RecordIOReader currentReader`), it  should be considered as without state to store.
As in the actual form of the PR, we are treating the symptom and not the cause, however in general I was wondering if the information of such state has to be contained in the DLQ input plugin or should be part of the `DeadLetterQueueReader` itself.

We can proceed in two steps:
- treat the symptom with this PR, that would fix actual installations (LS >= 8.4)
- fix the cause, so work on the LS core part

Fixing the cause would open a big question in my opinion. That would require a method (`DeadLetterQueueReader.setCurrentReaderAndPosition`) to return a value, this is a change that should be binary compatible, so it should work also with previous version of the plugin. Once we have a release of LS with that change we can proceed in updating then the DLQ input plugin. 
Having the DLQ input plugin as part of the LS core would greatly simplify the release process, should we consider moving it?

**Comment by kaisecheng** [comment]
Agree that we can proceed with this PR first and fix the state in core later.

Putting the DeadLetterQueueInputPlugin to core or plugin is about where to put the line between two. I think the line always exists, as the register(), close() and run() need to interact with the plugin. We need a better interface and move all `DeadLetterQueueReader` related state and SinceDB to core. So, when ruby plugin do close(), it only call the interface close without considering `queueReader`, `commitOffsets`, `readerHasState` and `sinceDb`

**Comment by jsvd** [comment]
> Having the DLQ input plugin as part of the LS core would greatly simplify the release process, should we consider moving it?

Fixing this in core touches on https://github.com/elastic/logstash/issues/8682. Plugins like this one, file input, jdbc and es input should be able to leverage storage primitives to store plugin state.

**Review by andsel — APPROVED** [review]
LGTM

**Pull request merged**